Add Stacked Bar Chart and sorting to Plots Tab

SURAJ-SHARMA27 opened the pull request:
Describe changes proposed in this pull request:
Addition of Sort By MinorCategory Options:

- Implemented a new feature allowing users to sort by minorCategory.
- Added a dropdown menu with minorCategory options for sorting.

Sorted Stacks by Selected MinorCategory:

- For the selected option from the dropdown, entities will appear at the beginning of each stack in a sorted order with respect to the count of chosen minorCategory.


## Any screenshots or GIFs?
[screen-capture (9).webm](https://github.com/user-attachments/assets/fc1d29a4-fc8a-4980-9444-25eae8a9e2a0)


## Notify reviewers
@alisman 
@sowmiyaa-kumar
@zeynepkaragoz
@TJMKuijpers
@inodb

inodb commented:
@SURAJ-SHARMA27 this looks awesome - thank you!

For the sorting, can we also sort by "# samples"? The value of the Y-axis basically

<img width="312" alt="image" src="https://github.com/user-attachments/assets/c2e002ed-7b1c-4813-bd3d-5dedf3419280">

This would be similar to e.g. how it works here for [treatment response](https://www.cbioportal.org/study/plots?id=ccle_broad_2019&plots_horz_selection=%7B%22dataType%22%3A%22none%22%7D&plots_vert_selection=%7B%22dataType%22%3A%22TREATMENT_RESPONSE%22%7D&plots_coloring_selection=%7B%7D):
<img width="328" alt="image" src="https://github.com/user-attachments/assets/c4f131c3-269a-418c-9e96-59b9323a886b">


This is slightly separate, but maybe for treatment response, the sorting should be using your new "Sort By" component too

schultzn commented:
Looks great. It would be nice to also be able to sort by the overall number of samples (I see that comment above) and also to change the sort order back to alphabetical.

jjgao commented:
Good work @SURAJ-SHARMA27 !

- Somehow x-axis labels does not change - always alphabetically, but the bars are changing

![image](https://github.com/user-attachments/assets/0662bf20-83e5-45e9-becc-3f3da59242a6)

- When switch Plot Type, should we change the sort value, ie. sort by number of samples if it's "Stacked bar chart" and by % samples if it's "100% stacked bar chart"?

- Should the 'Sort By' menu be included in the 'Vertical Axis' section since it's using the values from vertical axis.

SURAJ-SHARMA27 commented:
> Good work @SURAJ-SHARMA27 !
> 
> * Somehow x-axis labels does not change - always alphabetically, but the bars are changing
> 
> ![image](https://private-user-images.githubusercontent.com/840895/355878182-0662bf20-83e5-45e9-becc-3f3da59242a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNDQ5NDEsIm5iZiI6MTcyMzA0NDY0MSwicGF0aCI6Ii84NDA4OTUvMzU1ODc4MTgyLTA2NjJiZjIwLTgzZTUtNDVlOS1iZWNjLTNmM2RhNTkyNDJhNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QxNTMwNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZjkwMTRkOWNlY2M3NjIyMjBiYTZjNDE4MjcyN2FlOGJiNjNiMDI0ZWJlN2I0OTNlYzQ1ZmQ4MDg1MDE0NmI2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CuvfPWvGXjCrD_bwDyi2sMZAUnJLLuqxNoHECgpqC8Y)
> 
> * When switch Plot Type, should we change the sort value, ie. sort by number of samples if it's "Stacked bar chart" and by % samples if it's "100% stacked bar chart"?
> * Should the 'Sort By' menu be included in the 'Vertical Axis' section since it's using the values from vertical axis.
> * It would be nice to add the sort by parameter to the url.

Actually, I implemented that earlier, as you can see in the video I attached at the time of the PR. The labels were changing, but I forgot to include the logic when pushing it. Thank you for pointing that out. I have now made the changes and pushed it.

SURAJ-SHARMA27 commented:
> @SURAJ-SHARMA27 this looks awesome - thank you!
> 
> For the sorting, can we also sort by "# samples"? The value of the Y-axis basically
> 
> <img alt="image" width="312" src="https://private-user-images.githubusercontent.com/1334004/355856851-c2e002ed-7b1c-4813-bd3d-5dedf3419280.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNDc5NjksIm5iZiI6MTcyMzA0NzY2OSwicGF0aCI6Ii8xMzM0MDA0LzM1NTg1Njg1MS1jMmUwMDJlZC03YjFjLTQ4MTMtYmQzZC01ZGVkZjM0MTkyODAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDdUMTYyMTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY3ZTVmY2M1ZTJmNTM1NjhmYTVlNDdiMjIzMjg5Zjg0MGQ5ZWQ3MzViMzJjYjczYjdlN2E4MTNmMjUxOTA4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.HJ7qaTGOcU4Ojkz-n-JdL4FLC1ZYTedIWbn0QFRpFEg">
> This would be similar to e.g. how it works here for [treatment response](https://www.cbioportal.org/study/plots?id=ccle_broad_2019&plots_horz_selection=%7B%22dataType%22%3A%22none%22%7D&plots_vert_selection=%7B%22dataType%22%3A%22TREATMENT_RESPONSE%22%7D&plots_coloring_selection=%7B%7D): <img alt="image" width="328" src="https://private-user-images.githubusercontent.com/1334004/355857673-c4f131c3-269a-418c-9e96-59b9323a886b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNDc5NjksIm5iZiI6MTcyMzA0NzY2OSwicGF0aCI6Ii8xMzM0MDA0LzM1NTg1NzY3My1jNGYxMzFjMy0yNjlhLTQxOGMtOWU5Ni01OWI5MzIzYTg4NmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDdUMTYyMTA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWRkNmMxNjBlZWVkN2UwNmI3M2Y2ZDUzNGFkNThlZGRiOGNiY2M5NDZkMDUxNDIzMGMyY2I1ZDA5OGRmYTEzNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VqFxgZcGkRj6abvncrTCmFCwy1PQDKNCI5zd_gH32Sk">
> 
> This is slightly separate, but maybe for treatment response, the sorting should be using your new "Sort By" component too

Sir, I just want to be sure before starting the work: should the feature  include sorting based on the total number of samples per bar? This means the longest bar should come first, followed by the others accordingly.

SURAJ-SHARMA27 commented:
I think I have implemented most of the suggestions and added two more options for sorting: sort by the number of samples and alphabetically. If you could review it. @inodb 

[screen-capture (14).webm](https://github.com/user-attachments/assets/67779d29-2ab3-477d-9a33-73f075b58c11)

inodb commented:
Amazing - nice work @SURAJ-SHARMA27 !

> Should the 'Sort By' menu be included in the 'Vertical Axis' section since it's using the values from vertical axis.

@jjgao Good question - I kinda like it separate from x/y axis selection since we might also want to allow sorting by multiple values in the future (which could be like a multi-select dropdown element)

Should we rename `# samples` to `Number of Samples`? I like the consistency with the current y-axis and we use `#` in several places, but it's maybe not as obvious in the sorting what `#` means

gblaih commented on src/pages/groupComparison/MultipleCategoryBarPlot.tsx:
```diff
@@ -58,6 +58,11 @@ export interface IMultipleCategoryBarPlotProps {
     svgRef?: (svgContainer: SVGElement | null) => void;
     pValue: number | null;
     qValue: number | null;
+    SortByDropDownOptions?: { value: string; label: string }[];
```
sortByDropDownOptions

gblaih commented on src/pages/groupComparison/MultipleCategoryBarPlot.tsx:
```diff
@@ -425,6 +430,43 @@ export default class MultipleCategoryBarPlot extends React.Component<
 
     @computed get labels() {
         if (this.data.length > 0) {
+            if (this.props.sortByOption == 'SortByTotalSum') {
```
===

gblaih commented on src/pages/groupComparison/MultipleCategoryBarPlot.tsx:
```diff
@@ -425,6 +430,43 @@ export default class MultipleCategoryBarPlot extends React.Component<
+                );
+                return sortedMajorCategories;
+            } else if (
+                this.props.sortByOption != '' &&
```
lets use === and !== throughout the code

gblaih commented on src/shared/components/plots/PlotsTab.tsx:
```diff
@@ -5400,6 +5432,13 @@ export default class PlotsTab extends React.Component<IPlotsTabProps, {}> {
                                     percentage={isPercentage}
                                     stacked={isStacked}
+                                    SortByDropDownOptions={
+                                        this?.SortByDropDownOptions
+                                    }
+                                    updateDropDownOptions={
+                                        this?.updateDropDownOptions
+                                    }
+                                    sortByOption={this?.sortByOption}
```
is `this?.` necessary? i'd think `this.` is fine

gblaih commented on src/pages/groupComparison/MultipleCategoryBarPlot.tsx:
```diff
@@ -425,6 +430,43 @@ export default class MultipleCategoryBarPlot extends React.Component<
 
     @computed get labels() {
         if (this.data.length > 0) {
+            if (this.props.sortByOption == 'SortByTotalSum') {
+                const majorCategoryCounts: any = {};
+
+                this.data.forEach(item => {
```
this code looks pretty similar to the the function `sortDataByOption` you made below. it would be nice to extract this into a function that can be reused

gblaih commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,72 @@ export function sortDataByCategory<D>(
+            (a, b) => majorCategoryCounts[b] - majorCategoryCounts[a]
+        );
+
+        const reorderCounts = (counts: any) => {
```
instead of making this function inside, lets extract this into a new function outside for clarity and reuse

SURAJ-SHARMA27 commented:
I have implemented all the changes as suggested. You can review them. I am also attaching a video for reference.
@inodb  @gblaih 

[screen-capture (15).webm](https://github.com/user-attachments/assets/22665468-8985-4293-829c-df336e717a1c)

alisman commented on src/pages/groupComparison/MultipleCategoryBarPlot.tsx:
```diff
@@ -435,6 +451,18 @@ export default class MultipleCategoryBarPlot extends React.Component<
         }
     }
 
+    private setInitialSelectedOption = () => {
```
this is a bit of an anti-pattern.  calling from the child component to update the state of parent component on render.  it should be possible to do this in parent component prior to mount?

SURAJ-SHARMA27 replied:
In the parent component, it doesn't contain data in the form of `IMultipleCategoryBarPlotData`, from which I have to extract all the unique major categories for dropdown options. Therefore, to avoid extra calculations, I passed the prop from the parent and updated it when it became available in the child.

alisman commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,71 @@ export function sortDataByCategory<D>(
         }
     });
 }
+export function getSortedMajorCategories(
```
can we have a comment explaining what is meant by "Major Category" ?

SURAJ-SHARMA27 replied:
![image](https://github.com/user-attachments/assets/8928818d-ccfd-4ef3-b8b9-c2ad71563444)


The `data `is defined in a way that it contains x-axis labels referred to as major categories. This functionality is used twice: once during the sorting of data and again when the getLabels function in MultipleCategoryBarPlot.tsx is called. Therefore, I created a separate function for that to enhance reusability.

alisman commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,71 @@ export function sortDataByCategory<D>(
+    if (sortByOption === 'SortByTotalSum') {
+        const majorCategoryCounts: any = {};
+
+        data.forEach(item => {
```
i think you can do this operation with flatMap, which will avoid nesting forEach (which is confusing)

SURAJ-SHARMA27 replied:
The above function is used to find the sum of each `majorCategory` for different `minorCategories`. For instance, consider the following example:

data contains three `minorCategory` names: Biopsy, Resection, and Cusa. Each of these `minorCategories` contains a large array named counts, and each element in that array is an object containing `majorCategory`, count, and percentage. For example, for Biopsy, the `majorCategories` might include Germ Cell Tumor and Thyroid Cancer....

I am maintaining a totalSum of these majorCategories and then applying sorting.

If I use flatMap, it would first be difficult to determine which majorCategory belongs to which minorCategory. Additionally, I think it would require more space and have an equivalent time complexity of O(n * m) as it does now with the two nested loops. First, I would need to perform the flatMap operation (a map operation followed by a flattening operation, essentially equivalent to a nested loop) and then use reduce to accumulate the counts. What are your suggestions should i follow another approach?

alisman commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,71 @@ export function sortDataByCategory<D>(
+
+            return sortedEntityData.counts.map(item => item.majorCategory);
+        }
+    }
```
just complete the else here.

SURAJ-SHARMA27 replied:
yes completed

alisman commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,71 @@ export function sortDataByCategory<D>(
+    data: IMultipleCategoryBarPlotData[],
+    sortByOption: string | undefined
+): string[] {
+    if (sortByOption === 'SortByTotalSum') {
```
see if SortByTotalSum is available on an enum somewhere.  seems likely that it is

SURAJ-SHARMA27 replied:
No, it was not there; I have defined an enumeration and used it

alisman commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,71 @@ export function sortDataByCategory<D>(
+        return Object.keys(majorCategoryCounts).sort(
+            (a, b) => majorCategoryCounts[b] - majorCategoryCounts[a]
+        );
+    } else if (sortByOption !== '' && sortByOption !== 'alphabetically') {
```
use constant for 'alphabetically' or see if it's available on an enum

SURAJ-SHARMA27 replied:
I have defined an enumeration and used it

alisman commented on src/shared/components/plots/MultipleCategoryBarPlotUtils.ts:
```diff
@@ -98,6 +98,71 @@ export function sortDataByCategory<D>(
+    const sortedMajorCategories = getSortedMajorCategories(data, sortByOption);
+
+    if (sortByOption === 'SortByTotalSum' || sortedMajorCategories.length > 0) {
+        const reorderCounts = (counts: any) => {
```
avoid 'any'

SURAJ-SHARMA27 replied:
Yes, I have defined the types across all the changes.

inodb commented:
@SURAJ-SHARMA27 Just discussed this during the community call, some feedback from the group:

- When switching to percentage, the bars should sort by the percentage value (rather than absolute value)
- On the Table: maybe disable the sort here? Or if it is easy to enable sorting there too, enable it.
- Remove "Sort By" on charts where it is not functional (e.g. continuous vs continuous). Currently, the "Sort by" seems to show old values from a previous selection

SURAJ-SHARMA27 commented:
> @SURAJ-SHARMA27 Thanks for the fixes! I noticed the "Sort by" option is currently missing [here](https://deploy-preview-4960--cbioportalfrontend.netlify.app/study/plots?id=msk_impact_2017&plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_TYPE%22%7D&plots_vert_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22SPECIMEN_TYPE%22%7D&plots_coloring_selection=%7B%7D), do you see that?
> 
> ![image](https://private-user-images.githubusercontent.com/1334004/357450983-8eaef471-5031-447e-b11d-c4e3db3a7272.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NjIzMzgsIm5iZiI6MTcyMzU2MjAzOCwicGF0aCI6Ii8xMzM0MDA0LzM1NzQ1MDk4My04ZWFlZjQ3MS01MDMxLTQ0N2UtYjExZC1jNGUzZGIzYTcyNzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTNUMTUxMzU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmVhY2Q0Y2QxMjJiNzhkNDY4MWI3NzZiODU2OTVhN2I4ZTFlNDU5YzMxY2VmY2U1YTVlMzkyYTAwZTZhZGQ0ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.i9e6yge24dlObRJxi86msmDiyecmWo3pm6La8B1qoiY)

yes i fixed it

SURAJ-SHARMA27 commented:
> @SURAJ-SHARMA27 Just discussed this during the community call, people liked it! Some more feedback from the group:
> 
> * When switching to percentage, the bars should sort by the percentage value (rather than absolute value)
> * On the Table: maybe disable the sort here? Or if it is easy to enable sorting there too, enable it.
> * Remove "Sort By" on charts where it is not functional (e.g. continuous vs continuous). Currently, the "Sort by" seems to show old values from a previous selection

Thank you @inodb  sir, just few more doubts
i) This means that when the number of samples is selected from the dropdown, the data should be sorted by the percentage value.
ii) Sorting will be shown only for the stacked bar chart; I have fixed that.

inodb commented:
> i) This means that when the number of samples is selected from the dropdown, the data should be sorted by the percentage value.
 @SURAJ-SHARMA27 correct

SURAJ-SHARMA27 commented:
> > i) This means that when the number of samples is selected from the dropdown, the data should be sorted by the percentage value.
> 
> @SURAJ-SHARMA27 correct
> 
> > ii) Sorting will be shown only for the stacked bar chart; I have fixed that.
> 
> Thanks so much for fixing!

okay sir, I think I did it correctly. You can review it (pushed already). The bars will be sorted according to the percentages of the selected minor category. If the number of samples is selected from the dropdown, the bars are sorted by the total absolute count for each stack. For alphabetical sorting, the bars will be sorted as usual.

[screen-capture (33).webm](https://github.com/user-attachments/assets/f343c5b9-dd55-46c5-8756-0e48ebb00a78)

alisman commented:
Hi @SURAJ-SHARMA27,
There is a regression error caused by this PR.  I realize GSOC is long over.  Do you have a few minutes to look at the issue?

If you [visit this link](https://www.cbioportal.org/results/comparison?Z_SCORE_THRESHOLD=2.0&cancer_study_id=coadread_tcga_pub&cancer_study_list=coadread_tcga_pub&case_set_id=coadread_tcga_pub_nonhypermut&comparison_selectedGroups=%5B%22Altered%20group%22%2C%22Unaltered%20group%22%2C%22KRAS%22%2C%22NRAS%22%5D&gene_list=KRAS%20NRAS%20BRAF&gene_set_choice=user-defined-list&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations&comparison_subtab=clinical
) with the PR active, you'll see the following problem:

![image](https://github.com/user-attachments/assets/41b4901c-bd8c-41fe-b632-09c360356cd2)

If you are busy with school or work, we understand and will find someone else to look into it.

dippindots commented:
Related ticket: https://github.com/cBioPortal/cbioportal/issues/10722

alisman commented:
@SURAJ-SHARMA27 just FYI, we have a dev @gblaih who is taking care of the last fixes.

alisman commented on src/shared/components/plots/MultipleCategoryBarPlot.tsx:
```diff
@@ -913,8 +944,12 @@ export default class MultipleCategoryBarPlot extends React.Component<
+    componentDidUpdate(prevProps: IMultipleCategoryBarPlotProps) {
         this.updateLegendWidth();
+
+        if (this.props.horizontalBars !== prevProps.horizontalBars) {
```
@gblaih should be better to use the horizontalBars value in the component "key" attribute.  This will force it to re-mount when that value changes and so you can avoid dirtiness like the above.  You might lose other state, but it's a cleaner approach.  Give it a try.

alisman replied:
and apologies if this was not your change

inodb reviewed (approved):
This looks great!

There's one minor issue where occasionally when you mess around with swapping axes you see stuff like this:

![image](https://github.com/user-attachments/assets/28a8a0fb-2c42-477f-8df7-4124192ea4ce)


But haven't be able to figure out the steps to reproduce it reliably so I think ok to merge it and see if someone else comes up with a good way to reproduce it post merge
